File: src/services/auth.service.js
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const userRepository = require('../repositories/user.repository');

function isEmailValido(email) {
  return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
}

async function register({ nome, email, senha, role = 'cliente' }) {
  if (!nome || !nome.trim()) {
    const error = new Error('Nome é obrigatório.');
    error.statusCode = 400;
    throw error;
  }

  if (!email || !email.trim()) {
    const error = new Error('E-mail é obrigatório.');
    error.statusCode = 400;
    throw error;
  }

  if (!isEmailValido(email)) {
    const error = new Error('E-mail inválido.');
    error.statusCode = 400;
    throw error;
  }

  if (!senha || senha.length < 6) {
    const error = new Error('A senha deve ter pelo menos 6 caracteres.');
    error.statusCode = 400;
    throw error;
  }

  const existingUser = await userRepository.findByEmail(email.trim().toLowerCase());

  if (existingUser) {
    const error = new Error('E-mail já cadastrado.');
    error.statusCode = 409;
    throw error;
  }

  const senhaHash = await bcrypt.hash(senha, 10);

  const user = await userRepository.create({
    nome: nome.trim(),
    email: email.trim().toLowerCase(),
    senhaHash,
    role
  });

  return user;
}

async function login({ email, senha }) {
  if (!email || !senha) {
    const error = new Error('E-mail e senha são obrigatórios.');
    error.statusCode = 400;
    throw error;
  }

  const user = await userRepository.findByEmail(email.trim().toLowerCase());

  if (!user) {
    const error = new Error('Credenciais inválidas.');
    error.statusCode = 401;
    throw error;
  }

  const senhaValida = await bcrypt.compare(senha, user.senha_hash);

  if (!senhaValida) {
    const error = new Error('Credenciais inválidas.');
    error.statusCode = 401;
    throw error;
  }

  const token = jwt.sign(
    {
      id: user.id,
      nome: user.nome,
      email: user.email,
      role: user.role
    },
    process.env.JWT_SECRET,
    { expiresIn: process.env.JWT_EXPIRES_IN || '1d' }
  );

  return {
    token,
    user: {
      id: user.id,
      nome: user.nome,
      email: user.email,
      role: user.role
    }
  };
}

module.exports = {
  register,
  login
};

File: Backend/barberproapi/src/main/java/com/barberpro/barberproapi/service/AuthService.java
package com.barberpro.barberproapi.service;

import com.barberpro.barberproapi.domain.User;
import com.barberpro.barberproapi.domain.UserRole;
import com.barberpro.barberproapi.dto.LoginRequest;
import com.barberpro.barberproapi.dto.LoginResponse;
import com.barberpro.barberproapi.dto.RegisterRequest;
import com.barberpro.barberproapi.dto.UserResponse;
import com.barberpro.barberproapi.exception.BusinessException;
import com.barberpro.barberproapi.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;

    @Transactional
    public UserResponse register(RegisterRequest request) {
        String email = request.email().trim().toLowerCase();

        if (userRepository.existsByEmail(email)) {
            throw BusinessException.conflict("E-mail já cadastrado.");
        }

        User user = User.builder()
                .nome(request.nome().trim())
                .email(email)
                .senhaHash(passwordEncoder.encode(request.senha()))
                .role(UserRole.CLIENTE)
                .build();

        return UserResponse.from(userRepository.save(user));
    }

    public LoginResponse login(LoginRequest request) {
        String email = request.email().trim().toLowerCase();

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> BusinessException.unauthorized("Credenciais inválidas."));

        if (!passwordEncoder.matches(request.senha(), user.getSenhaHash())) {
            throw BusinessException.unauthorized("Credenciais inválidas.");
        }

        String token = jwtService.generateToken(user);
        return new LoginResponse(token, UserResponse.from(user));
    }
}
